Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,311 @@
+/-
+Copyright (c) 2023 Newell Jensen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Newell Jensen
+-/
+import Mathlib.Data.ENat.Basic
+import Mathlib.Data.Matrix.Notation
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.GroupTheory.PresentedGroup
+import Mathlib.LinearAlgebra.Matrix.Symmetric
+
+/-!
+# Coxeter Systems
+
+This file defines Coxeter systems.
+
+It introduces a type class `IsCoxeterGroup` for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+The finite Coxeter groups are classified as the four infinite families:
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional systems:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries
+  equal to one and off-diagonal entries distinct from one.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup B M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter System.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ)
+
+/-- A matrix `IsCoxeter` if it is a symmetric matrix with diagonal entries equal to one
+and off-diagonal entries distinct from one. -/
+class Matrix.IsCoxeter (M : Matrix B B ℕ) : Prop where
+  symmetric : M.IsSymm := by aesop
+  diagonal : ∀ i : B, M i i = 1 := by aesop
+  off_diagonal : ∀ i j : B, i ≠ j → M i j ≠ 1 := by aesop
+
+namespace CoxeterGroup
+
+namespace Relations
+
+/-- The relation terms corresponding to a Coxeter matrix. -/
+def ofMatrix : B × B → FreeGroup B :=
```

Comment:
The assumption `[M.IsCoxeter]` should be made here, or at the very least for the definition of `Matrix.CoxeterGroup`, because we do not want to consider the presented group `M.CoxeterGroup` unless we know that `M` satisfies the expected conditions.

Reply:
When I tried adding it to those definitions in different combinations, non of the them would pass the linter.  I was able to get it to pass the linter when adding `[M.IsCoxeter]` to both `CoxeterSystem` and `IsCoxeterGroup`.  Although this isn't as ideal as making the constraint on `Matrix.CoxeterGroup` it was the best that I was able to figure out where at least `CoxeterSystem` and `IsCoxeterGroup` both need the matrix to be a Coxeter matrix.

Do you know of another way of doing it better than this by adding `[M.IsCoxeter]` to the defintions somehow which would get past the linter?  After the code compiles and I save the file I do `lake build && lake exe runMathlibLinter Mathlib.GroupTheory.SpecificGroups.Coxeter` to run the linter.

Reply:
This is a case where it is reasonable to circumvent the `unusedArguments` linter. I think you can do
```lean
@[nolint unusedArguments]
def ofMatrix [M.IsCoxeter] : B × B → FreeGroup B :=
```
and similarly for the other definitions which may require it.